File: Data Structures & Algorithms/permutation-string/submission-0.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        sp = dict()
        for char in s1:
            sp[char] = sp.get(char, 0) + 1
        
        spCopy = dict(sp)
        total = len(s1)

        left = 0
        while left <= len(s2) - len(s1):
            right = left + len(s1) - 1
            i = left
            while i <= right:
                if s2[i] in spCopy:
                    if spCopy[s2[i]] > 0:
                        spCopy[s2[i]] -= 1
                        total -= 1
                    if total == 0:
                        return True
                else:
                    break
                i += 1

            spCopy = dict(sp)
            total = len(s1)
            left += 1

        return False

File: Data Structures & Algorithms/palindrome-partitioning/submission-0.java
class Solution {
    public List<List<String>> partition(String s) {
        return partitionHelper(s);
    }

    private List<List<String>> partitionHelper(String s) {
        List<List<String>> result = new ArrayList<>();
        if (s.isEmpty())
            return List.of(List.of());

        for (int i = 0; i < s.length(); i++) {
            if (isPalindrome(s, 0, i)) {
                String left = s.substring(0, i + 1);
                List<List<String>> right = partitionHelper(s.substring(i + 1));
                for (List<String> l : right) {
                    List<String> copy = new ArrayList<>(l.size() + 1);
                    copy.add(left);
                    copy.addAll(l);
                    result.add(copy);
                }

            }
        }

        return result;
    }

    private boolean isPalindrome(String s, int start, int end) {
        int left = start;
        int right = end;

        while (left <= end) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }

            left++;
            right--;
        }

        return true;
    }
}
